### PR TITLE
[Index] Avoid forming relations to non-indexed decls

### PR DESCRIPTION
**Pull request by hamishknight**
Make sure to check whether we are able to index a decl before reporting it as a relation. Additionally for containers, walk up the stack to find the first container that we can index, avoiding treating local variables as containers when we're not indexing them.

rdar://126137541

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/Index/Index.cpp [hunk 1]
```diff
@@ -216,26 +216,38 @@ class ContainerTracker {
...
+        return;
 
-    const StackEntry &Entry = Stack.back();
+      auto MapEntry = Entry.Containers.find(Entry.ActiveKey);
+      if (MapEntry == Entry.Containers.end())
+        return;
 
-    if (!Entry.ActiveKey)
-      return;
+      SmallVector<const Decl *, 4> Containers;
```
Comment (bnbarham) [inline lib/Index/Index.cpp:230]: Could have a bool for whether we actually allowed any decls or not and then just use `f(D)` in the two if's, but if you prefer this I'd move the `Containers` out of the loop.

**Review by bnbarham — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/Index/Index.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/Index/Index.cpp]: Sure, don't feel too strongly, switched to using a bool

**Pull request merged**